doc/analyzers/VSTHRD003.md

```diff
@@ -90,3 +92,176 @@ class SomeClass
+
+In the next example, a `TaskCompletionSource<T>` is used as a black-box for unblocking functionality.
+It too represents awaiting a foreign task:
+
+```cs
+class SomeClass
+{
+    TaskCompletionSource<bool> tcs = new();
+
+    public async Task MyMethodAsync()
+    {
+        await this.tcs.Task; /* This analyzer will report warning on this line. */
+        /* do more stuff */
+    }
+
+    void UnlockProgress()
+    {
+        this.tcs.TrySetResult(true);
+    }
+}
+```
+
+The problem with the above code is that `MyMethodAsync()` waits for unknown work (whatever work will lead to the completion of the `TaskCompletionSource`) before making progress.
+If `UnlockProgress()` is never called, the caller of `MyMethodAsync()` will be awaiting forever.
+Now suppose that the caller of `MyMethodAsync()` is actually inside a `JoinableTaskFactory.Run` delegate:
+
+```cs
+void SomeCaller()
+{
+    joinableTaskFactory.Run(async delegate
+    {
+        await someClass.MyMethodAsync();
+    });
+}
+```
+
+If `SomeCaller()` runs on the main thread, then it will effectively block the main thread while waiting for `this.tcs.Task` from `SomeClass` to complete.
+Now suppose that another thread comes along and wants to do some work before calling `UnlockProgress()`:
+
+```cs
+partial class SomeClass
+{
+    async Task KeyMasterAsync()
+    {
+        await joinableTaskFactory.SwitchToMainThreadAsync();
+        // do some work
+        // Unblock others
+        someClass.UnlockProgress();
+    }
+}
+```
+
+We have a deadlock, because `SomeCaller()` is blocking the main thread while waiting for `UnlockProgress()` to be called, but `UnlockProgress()` will not be called until `KeyMasterAsync` can reach the main thread.
+
+Fixing this fundamentally means that `SomeCaller` will need to *join* whatever work may be needed to ultimately call `UnlockProgress`. But for `SomeCaller`, that work is unknown, since it's at least partially inside another class.
+`TaskCompletionSource<T>` is fundamentally a blackbox and the most difficult thing to use correctly while avoiding deadlocks.
+
+Preferred solutions involve replacing `TaskCompletionSource<T>` with another type that makes tracking the work involved automatic.
+These include:
+
+1. Use `JoinableTaskFactory.RunAsync` and store the resulting `JoinableTask` in a field to await later.
+1. Use `AsyncLazy<T>` for one-time init work that should only start if required. Be sure to pass in a `JoinableTaskFactory` instance to its constructor.
+
+Assuming you must keep using `TaskCompletionSource<T>` though, here's how it can be done as safely as possible.
+Joining a set of unknown work is best done with the `JoinableTaskCollection` class.
+It is the responsibility of `SomeClass` in the example above to work with this collection to avoid deadlocks, like this:
+
+```cs
+class SomeClass
+{
+    TaskCompletionSource<bool> tcs = new();
+    JoinableTaskCollection jtc;
+    JoinableTaskFactory jtf;
+
+    SomeClass()
```

Comment:
```suggestion
    public SomeClass(JoinableTaskContext joinableTaskContext)
```

Reply:
I see below that you mention "Notice how the public API of the class does not need to expose any `JoinableTask`-related types.". It's not clear how this context gets here, nor the JTF below in `KeyMasterAsync`. Maybe that's fine.

Reply:
*How* `JoinableTaskContext` gets to the constructor is outside the scope of the sample, so I left it out. But I can see how it creates confusion. So I've added it to the constructor like you suggested, but the point is that the constructor is *not* public, as it may not need to be. Typically such a class as this is a 'service' rather than a data object, which is why it's tracking tasks long-term. And as a service, the ctor is typically _not_ public, and that's how I can say "see how the public API doesn't use JTF types?"

`KeyMasterAsync` just uses the `jtf` field. I'll add `this.` in front to make this more clear. Then the `joinableTaskFactory` variable that it just imagines up is, per the code comment above it, _any_ jtf it can get its hands on -- including the field. But I wanted to illustrate that it doesn't have to be the field.